File: app/src/main/java/com/dpis/module/templates/BatchScopeRequestCoordinator.java
package com.dpis.module.templates;

import com.dpis.module.BuildConfig;
import com.dpis.module.DpisApplication;
import com.dpis.module.R;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashSet;
import java.util.List;

import io.github.libxposed.service.XposedService;

public final class BatchScopeRequestCoordinator {
    public interface Host {
        void showToast(int messageResId, Object... formatArgs);

        void requestAppsLoad();

        void runOnUiThread(Runnable runnable);
    }

    interface ScopeRequester {
        List<String> getScope();

        void requestScope(List<String> packages, XposedService.OnScopeEventListener listener);
    }

    private final Host host;
    private final ScopeRequester scopeRequester;
    private final boolean modernFlavor;

    public BatchScopeRequestCoordinator(Host host) {
        this(host, fromService(DpisApplication.getXposedService()),
                "modern".equals(BuildConfig.FLAVOR));
    }

    BatchScopeRequestCoordinator(Host host, ScopeRequester scopeRequester, boolean modernFlavor) {
        this.host = host;
        this.scopeRequester = scopeRequester;
        this.modernFlavor = modernFlavor;
    }

    public Result requestMissingScope(List<String> successfulPackages) {
        LinkedHashSet<String> packages = sanitizePackages(successfulPackages);
        if (packages.isEmpty()) {
            return Result.noRequest();
        }
        if (!modernFlavor || scopeRequester == null) {
            notifyManualScopeRequired();
            return Result.manualRequired(packages.size());
        }

        List<String> scope;
        try {
            scope = scopeRequester.getScope();
        } catch (RuntimeException exception) {
            notifyManualScopeRequired();
            return Result.manualRequired(packages.size());
        }
        if (scope == null) {
            notifyManualScopeRequired();
            return Result.manualRequired(packages.size());
        }
        LinkedHashSet<String> missingPackages = new LinkedHashSet<>(packages);
        missingPackages.removeAll(new LinkedHashSet<>(scope));
        if (missingPackages.isEmpty()) {
            return Result.noRequest();
        }

        ArrayList<String> requestPackages = new ArrayList<>(missingPackages);
        try {
            scopeRequester.requestScope(requestPackages, new XposedService.OnScopeEventListener() {
                @Override
                public void onScopeRequestApproved(List<String> approved) {
                    if (host != null) {
                        host.runOnUiThread(() -> {
                            host.showToast(R.string.quick_template_scope_request_approved,
                                    approved != null ? approved.size() : 0);
                            host.requestAppsLoad();
                        });
                    }
                }

                @Override
                public void onScopeRequestFailed(String message) {
                    if (host != null) {
                        host.runOnUiThread(() -> {
                            host.showToast(R.string.quick_template_scope_manual_required);
                            host.requestAppsLoad();
                        });
                    }
                }
            });
            if (host != null) {
                host.showToast(R.string.quick_template_scope_request_started,
                        requestPackages.size());
            }
            return Result.requestStarted(requestPackages);
        } catch (RuntimeException exception) {
            notifyManualScopeRequired();
            return Result.manualRequired(requestPackages.size());
        }
    }

    private void notifyManualScopeRequired() {
        if (host != null) {
            host.showToast(R.string.quick_template_scope_manual_required);
        }
    }

    private static ScopeRequester fromService(XposedService service) {
        if (service == null) {
            return null;
        }
        return new ScopeRequester() {
            @Override
            public List<String> getScope() {
                return service.getScope();
            }

            @Override
            public void requestScope(List<String> packages,
                    XposedService.OnScopeEventListener listener) {
                service.requestScope(packages, listener);
            }
        };
    }

    private static LinkedHashSet<String> sanitizePackages(List<String> packages) {
        LinkedHashSet<String> sanitized = new LinkedHashSet<>();
        if (packages == null) {
            return sanitized;
        }
        for (String packageName : packages) {
            if (packageName == null) {
                continue;
            }
            String trimmed = packageName.trim();
            if (!trimmed.isEmpty()) {
                sanitized.add(trimmed);
            }
        }
        return sanitized;
    }

    public static final class Result {
        final boolean requestStarted;
        final boolean manualRequired;
        final List<String> requestedPackages;
        final int affectedPackageCount;

        private Result(boolean requestStarted,
                boolean manualRequired,
                List<String> requestedPackages,
                int affectedPackageCount) {
            this.requestStarted = requestStarted;
            this.manualRequired = manualRequired;
            this.requestedPackages = Collections.unmodifiableList(new ArrayList<>(
                    requestedPackages != null ? requestedPackages : Collections.emptyList()));
            this.affectedPackageCount = affectedPackageCount;
        }

        static Result noRequest() {
            return new Result(false, false, Collections.emptyList(), 0);
        }

        static Result requestStarted(List<String> requestedPackages) {
            return new Result(true, false, requestedPackages,
                    requestedPackages != null ? requestedPackages.size() : 0);
        }

        static Result manualRequired(int affectedPackageCount) {
            return new Result(false, true, Collections.emptyList(), affectedPackageCount);
        }
    }
}

File: app/src/main/java/com/dpis/module/templates/BatchScopeRequestCoordinator.kt
package com.dpis.module.templates

import com.dpis.module.BuildConfig
import com.dpis.module.DpisApplication
import com.dpis.module.R
import com.dpis.module.settings.ScopeRequestGate
import io.github.libxposed.service.XposedService

class BatchScopeRequestCoordinator internal constructor(
    private val host: Host?,
    private val scopeRequester: ScopeRequester?,
    private val modernFlavor: Boolean,
) {
    constructor(host: Host?) : this(
        host,
        fromService(DpisApplication.xposedService),
        BuildConfig.FLAVOR == "modern",
    )
    interface Host {
        fun showToast(messageResId: Int, vararg formatArgs: Any?)

        fun requestAppsLoad()

        fun runOnUiThread(runnable: Runnable)
    }

    internal interface ScopeRequester {
        fun getScope(): List<String>?

        fun requestScope(packages: List<String>, listener: XposedService.OnScopeEventListener)
    }

    fun requestMissingScope(successfulPackages: List<String>?): Result {
        val packages = successfulPackages.orEmpty()
            .asSequence()
            .map { it.trim() }
            .filter(String::isNotEmpty)
            .toCollection(LinkedHashSet())
        if (packages.isEmpty()) return Result.noRequest()
        val requester = scopeRequester
        if (!modernFlavor || requester == null) {
            notifyManualScopeRequired()
            return Result.manualRequired(packages.size)
        }

        val scope = try {
            requester.getScope()
        } catch (_: RuntimeException) {
            null
        }
        if (scope == null) {
            notifyManualScopeRequired()
            return Result.manualRequired(packages.size)
        }
        val requestPackages = (packages - scope.toSet()).toList()
        if (requestPackages.isEmpty()) return Result.noRequest()

        val request = ScopeRequestGate.shared().tryStart("template-batch", requestPackages)
            ?: run {
                host?.showToast(R.string.scope_request_pending)
                return Result.requestAlreadyPending(requestPackages.size)
            }
        return try {
            requester.requestScope(requestPackages, object : XposedService.OnScopeEventListener {
                override fun onScopeRequestApproved(approved: MutableList<String>) {
                    request.finish("approved")
                    val requestHost = host ?: return
                    requestHost.runOnUiThread(Runnable {
                        requestHost.showToast(
                            R.string.quick_template_scope_request_approved,
                            approved.size,
                        )
                        requestHost.requestAppsLoad()
                    })
                }

                override fun onScopeRequestFailed(message: String) {
                    request.finish("failed")
                    val requestHost = host ?: return
                    requestHost.runOnUiThread(Runnable {
                        requestHost.showToast(R.string.quick_template_scope_manual_required)
                        requestHost.requestAppsLoad()
                    })
                }
            })
            host?.showToast(R.string.quick_template_scope_request_started, requestPackages.size)
            Result.requestStarted(requestPackages)
        } catch (_: RuntimeException) {
            request.finish("exception")
            notifyManualScopeRequired()
            Result.manualRequired(requestPackages.size)
        }
    }

    private fun notifyManualScopeRequired() {
        host?.showToast(R.string.quick_template_scope_manual_required)
    }

    class Result private constructor(
        @JvmField val requestStarted: Boolean,
        @JvmField val manualRequired: Boolean,
        @JvmField val requestedPackages: List<String>,
        @JvmField val affectedPackageCount: Int,
    ) {
        companion object {
            fun noRequest() = Result(false, false, emptyList(), 0)

            fun requestStarted(requestedPackages: List<String>) = Result(
                true,
                false,
                requestedPackages.toList(),
                requestedPackages.size,
            )

            fun manualRequired(affectedPackageCount: Int) = Result(
                false,
                true,
                emptyList(),
                affectedPackageCount,
            )

            fun requestAlreadyPending(affectedPackageCount: Int) = Result(
                false,
                false,
                emptyList(),
                affectedPackageCount,
            )
        }
    }

    companion object {
        private fun fromService(service: XposedService?): ScopeRequester? = service?.let { target ->
            object : ScopeRequester {
                override fun getScope(): List<String> = target.scope

                override fun requestScope(
                    packages: List<String>,
                    listener: XposedService.OnScopeEventListener,
                ) = target.requestScope(packages, listener)
            }
        }
    }
}
